File: src/java/java5/src/hprose/common/HproseInvocationHandler.java
/**********************************************************\
|                                                          |
|                          hprose                          |
|                                                          |
| Official WebSite: http://www.hprose.com/                 |
|                   http://www.hprose.net/                 |
|                   http://www.hprose.org/                 |
|                                                          |
\**********************************************************/
/**********************************************************\
 *                                                        *
 * HproseInvocationHandler.java                           *
 *                                                        *
 * hprose InvocationHandler class for Java.               *
 *                                                        *
 * LastModified: May 6, 2011                              *
 * Author: Ma Bingyao <andot@hprose.com>                  *
 *                                                        *
\**********************************************************/
package hprose.common;

import java.lang.reflect.InvocationHandler;
import java.lang.reflect.Method;
import java.lang.reflect.Type;

public class HproseInvocationHandler implements InvocationHandler {
    private static final Byte byteZero = Byte.valueOf((byte) 0);
    private static final Short shortZero = Short.valueOf((short) 0);
    private static final Integer intZero = Integer.valueOf(0);
    private static final Long longZero = Long.valueOf((long) 0);
    private static final Character charZero = Character.valueOf((char) 0);
    private static final Float floatZero = new Float((float) 0);
    private static final Double doubleZero = new Double((double) 0);

    private HproseInvoker client;
    private String ns;

    public HproseInvocationHandler(HproseInvoker client, String ns) {
        this.client = client;
        this.ns = ns;
    }

    public Object invoke(Object proxy, Method method, Object[] arguments) throws Throwable {
        Class<?>[] paramTypes = method.getParameterTypes();
        Type returnType = method.getGenericReturnType();
        if (void.class.equals(returnType) ||
            Void.class.equals(returnType)) {
            returnType = null;
        }
        int n = paramTypes.length;
        String functionName = method.getName();
        if (ns != null) {
            functionName = ns + '_' + functionName;
        }
        Object result = null;
        if ((n > 0) && (paramTypes[n - 1].equals(HproseCallback.class))) {
            HproseCallback callback = (HproseCallback) arguments[n - 1];
            Object[] tmpargs = new Object[n - 1];
            System.arraycopy(arguments, 0, tmpargs, 0, n - 1);
            client.invoke(functionName, tmpargs, callback, returnType);
        }
        else if ((n > 1) &&
                 (paramTypes[n - 2].equals(HproseCallback.class)) &&
                 (paramTypes[n - 1].equals(boolean.class))) {
            HproseCallback callback = (HproseCallback) arguments[n - 2];
            boolean byRef = ((Boolean) arguments[n - 1]).booleanValue();
            Object[] tmpargs = new Object[n - 2];
            System.arraycopy(arguments, 0, tmpargs, 0, n - 2);
            client.invoke(functionName, tmpargs, callback, returnType, byRef);
        }
        else if ((n > 1) &&
                 (paramTypes[n - 2].equals(HproseCallback.class)) &&
                 (paramTypes[n - 1].equals(Class.class))) {
            HproseCallback callback = (HproseCallback) arguments[n - 2];
            returnType = (Class) arguments[n - 1];
            Object[] tmpargs = new Object[n - 2];
            System.arraycopy(arguments, 0, tmpargs, 0, n - 2);
            client.invoke(functionName, tmpargs, callback, returnType);
        }
        else if ((n > 2) &&
                 (paramTypes[n - 3].equals(HproseCallback.class)) &&
                 (paramTypes[n - 2].equals(Class.class)) &&
                 (paramTypes[n - 1].equals(boolean.class))) {
            HproseCallback callback = (HproseCallback) arguments[n - 3];
            returnType = (Class) arguments[n - 2];
            boolean byRef = ((Boolean) arguments[n - 1]).booleanValue();
            Object[] tmpargs = new Object[n - 3];
            System.arraycopy(arguments, 0, tmpargs, 0, n - 3);
            client.invoke(functionName, tmpargs, callback, returnType, byRef);
        }
        else {
            result = client.invoke(functionName, arguments, returnType);
            if (result instanceof HproseException) {
                throw (HproseException) result;
            }
        }
        if (result == null) {
            if (int.class.equals(returnType)) {
                return intZero;
            }
            if (long.class.equals(returnType)) {
                return longZero;
            }
            if (byte.class.equals(returnType)) {
                return byteZero;
            }
            if (short.class.equals(returnType)) {
                return shortZero;
            }
            if (float.class.equals(returnType)) {
                return floatZero;
            }
            if (double.class.equals(returnType)) {
                return doubleZero;
            }
            if (char.class.equals(returnType)) {
                return charZero;
            }
            if (boolean.class.equals(returnType)) {
                return Boolean.FALSE;
            }
        }
        return result;
    }
}

File: src/dotnet/src/Hprose/Client/HproseInvocationHandler.cs
/**********************************************************\
|                                                          |
|                          hprose                          |
|                                                          |
| Official WebSite: http://www.hprose.com/                 |
|                   http://www.hprose.net/                 |
|                   http://www.hprose.org/                 |
|                                                          |
\**********************************************************/
/**********************************************************\
 *                                                        *
 * HproseInvocationHandler.cs                             *
 *                                                        *
 * hprose InvocationHandler class for C#.                 *
 *                                                        *
 * LastModified: Feb 18, 2014                             *
 * Author: Ma Bingyao <andot@hprose.com>                  *
 *                                                        *
\**********************************************************/
#if !(PocketPC || Smartphone || WindowsCE || WINDOWS_PHONE || Core)
using System;
using System.Numerics;
using System.Reflection;
using System.Threading;
using Hprose.IO;
using Hprose.Reflection;
using Hprose.Common;

namespace Hprose.Client {
    class HproseInvocationHandler : IInvocationHandler {
        private String ns;
        private HproseClient client;

        public HproseInvocationHandler(HproseClient client, String ns) {
            this.client = client;
            this.ns = ns;
        }

        private static Type[] GetTypes(ParameterInfo[] parameters) {
            int n = parameters.Length;
            Type[] types = new Type[n];
            for (int i = 0; i < n; i++) {
                types[i] = parameters[i].ParameterType;
            }
            return types;
        }

        public object Invoke(object proxy, MethodInfo method, object[] args) {
            ParameterInfo[] parameters = method.GetParameters();
            Type[] paramTypes = GetTypes(parameters);
            Type returnType = method.ReturnType;
            if (returnType == typeof(void)) {
                returnType = null;
            }
            int n = paramTypes.Length;
            string functionName = method.Name;
            if (ns != null) {
                functionName = ns + '_' + functionName;
            }
            bool byRef = false;
            foreach (Type param in paramTypes) {
                if (param.IsByRef) {
                    byRef = true;
                    break;
                }
            }
            if ((n > 0) && (paramTypes[n - 1] == typeof(HproseCallback))) {
                HproseCallback callback = (HproseCallback)args[n - 1];
                object[] tmpargs = new object[n - 1];
                Array.Copy(args, 0, tmpargs, 0, n - 1);
                client.Invoke(functionName, tmpargs, callback, byRef);
                return null;
            }
            if ((n > 0) && (paramTypes[n - 1] == typeof(HproseCallback1))) {
                HproseCallback1 callback = (HproseCallback1)args[n - 1];
                object[] tmpargs = new object[n - 1];
                Array.Copy(args, 0, tmpargs, 0, n - 1);
                client.Invoke(functionName, tmpargs, callback);
                return null;
            }
            if ((n > 1) && (paramTypes[n - 2] == typeof(HproseCallback)) &&
                           (paramTypes[n - 1] == typeof(Type))) {
                HproseCallback callback = (HproseCallback)args[n - 2];
                returnType = (Type)args[n - 1];
                object[] tmpargs = new object[n - 2];
                Array.Copy(args, 0, tmpargs, 0, n - 2);
                client.Invoke(functionName, tmpargs, callback, returnType, byRef);
                return null;
            }
            if ((n > 1) && (paramTypes[n - 2] == typeof(HproseCallback1)) &&
                           (paramTypes[n - 1] == typeof(Type))) {
                HproseCallback1 callback = (HproseCallback1)args[n - 2];
                returnType = (Type)args[n - 1];
                object[] tmpargs = new object[n - 2];
                Array.Copy(args, 0, tmpargs, 0, n - 2);
                client.Invoke(functionName, tmpargs, callback, returnType);
                return null;
            }
#if !(dotNET10 || dotNET11 || dotNETCF10)
            if ((n > 0) && (paramTypes[n - 1].IsGenericType &&
                           (paramTypes[n - 1].GetGenericTypeDefinition() == typeof(HproseCallback<>)))) {
                IInvokeHelper helper = Activator.CreateInstance(typeof(InvokeHelper<>).MakeGenericType(paramTypes[n - 1].GetGenericArguments())) as IInvokeHelper;
                Delegate callback = (Delegate)args[n - 1];
                object[] tmpargs = new object[n - 1];
                Array.Copy(args, 0, tmpargs, 0, n - 1);
                helper.Invoke(client, functionName, tmpargs, callback, byRef);
                return null;
            }
            if ((n > 0) && (paramTypes[n - 1].IsGenericType &&
                           (paramTypes[n - 1].GetGenericTypeDefinition() == typeof(HproseCallback1<>)))) {
                IInvokeHelper1 helper = Activator.CreateInstance(typeof(InvokeHelper1<>).MakeGenericType(paramTypes[n - 1].GetGenericArguments())) as IInvokeHelper1;
                Delegate callback = (Delegate)args[n - 1];
                object[] tmpargs = new object[n - 1];
                Array.Copy(args, 0, tmpargs, 0, n - 1);
                helper.Invoke(client, functionName, tmpargs, callback);
                return null;
            }
#endif
            object result = client.Invoke(functionName, args, returnType, byRef);
            if (result is Exception) {
                throw (Exception)result;
            }
            return result;
        }
    }
}
#endif